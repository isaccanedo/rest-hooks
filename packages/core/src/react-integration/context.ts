import { createContext } from 'react';
import type { DenormalizeCache } from '@rest-hooks/normalizr';

import type { ActionTypes, State } from '../types.js';
import { initialState } from '../state/createReducer.js';
import Controller from '../controller/Controller.js';

export const StateContext = createContext(initialState);

const dispatch = (value: ActionTypes) => {
  /* istanbul ignore next */
  if (process.env.NODE_ENV !== 'production') {
    console.error(
      'It appears you are trying to use Rest Hooks without a provider.\nFollow instructions: https://resthooks.io/docs/getting-started/installation#add-provider-at-top-level-component',
    );
    /* istanbul ignore next */
    if (process.env.NODE_ENV === 'test') {
      console.error(
        'If you are trying to test: https://resthooks.io/docs/guides/unit-testing-hooks',
      );
    }
  }
  return Promise.resolve();
};
export const DispatchContext = createContext(dispatch);

// this is not needed anymore, but keeping around for backcompatibility
export const DenormalizeCacheContext = createContext<DenormalizeCache>({
  entities: {},
  results: {},
});

export const ControllerContext = createContext<Controller>(
  new Controller({
    dispatch,
    globalCache: {
      entities: {},
      results: {},
    },
  }),
);

export interface Store<S> {
  subscribe(listener: () => void): () => void;
  dispatch: React.Dispatch<ActionTypes>;
  getState(): S;
  uninitialized?: boolean;
}
/* istanbul ignore next */
export const StoreContext = createContext<Store<State<unknown>>>({
  subscribe: listener => () => {},
  dispatch: () => {},
  getState: () => initialState,
  uninitialized: true,
});
