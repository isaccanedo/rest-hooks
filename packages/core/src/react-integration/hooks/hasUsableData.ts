import { FetchShape } from '../../endpoint/index.js';

/** @deprecated use https://resthooks.io/docs/api/Controller#getResponse */
export default function hasUsableData(
  fetchShape: Pick<FetchShape<any>, 'options'>,
  cacheReady: boolean,
  deleted: boolean,
  invalidated?: boolean,
) {
  return (
    !deleted &&
    !fetchShape.options?.invalidIfStale &&
    cacheReady &&
    !invalidated
  );
}
