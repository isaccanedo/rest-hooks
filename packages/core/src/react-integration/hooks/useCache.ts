/* eslint-disable @typescript-eslint/ban-types */
import { useMemo } from 'react';
import {
  EndpointInterface,
  DenormalizeNullable,
  Schema,
  FetchFunction,
  ResolveType,
} from '@rest-hooks/normalizr';

import { ReadShape, ParamsFromShape } from '../../endpoint/index.js';
import shapeToEndpoint from '../../endpoint/adapter.js';
import useCacheNew from '../newhooks/useCache.js';

/**
 * Access a response if it is available.
 *
 * `useCache` guarantees referential equality globally.
 * @see https://resthooks.io/docs/api/useCache
 */
export default function useCache<
  E extends
    | Pick<
        EndpointInterface<FetchFunction, Schema | undefined, undefined>,
        'key' | 'schema' | 'invalidIfStale'
      >
    | Pick<ReadShape<any, any>, 'getFetchKey' | 'schema' | 'options'>,
  Args extends
    | (E extends { key: any }
        ? readonly [...Parameters<E['key']>]
        : readonly [ParamsFromShape<E>])
    | readonly [null],
>(
  endpoint: E,
  ...args: Args
): E['schema'] extends {}
  ? DenormalizeNullable<E['schema']>
  : E extends (...args: any) => any
  ? ResolveType<E> | undefined
  : any {
  const adaptedEndpoint: any = useMemo(() => {
    return shapeToEndpoint(endpoint);
    // we currently don't support shape changes
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return useCacheNew(adaptedEndpoint, ...args);
}
