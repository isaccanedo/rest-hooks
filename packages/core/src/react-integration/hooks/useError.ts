import type { NetworkError, UnknownError } from '@rest-hooks/normalizr';
import { useMemo } from 'react';
import {
  EndpointInterface,
  Schema,
  FetchFunction,
} from '@rest-hooks/normalizr';

import { ReadShape, ParamsFromShape } from '../../endpoint/index.js';
import shapeToEndpoint from '../../endpoint/adapter.js';
import useErrorNew from '../newhooks/useError.js';

export type ErrorTypes = NetworkError | UnknownError;

type UseErrorReturn<P> = P extends [null] ? undefined : ErrorTypes | undefined;

/**
 * Get any errors for a given request
 * @see https://resthooks.io/docs/api/useError
 */
export default function useError<
  E extends
    | Pick<
        EndpointInterface<FetchFunction, Schema | undefined, undefined>,
        'key' | 'schema' | 'invalidIfStale'
      >
    | Pick<ReadShape<any, any>, 'getFetchKey' | 'schema' | 'options'>,
  Args extends
    | (E extends { key: any }
        ? readonly [...Parameters<E['key']>]
        : readonly [ParamsFromShape<E>])
    | readonly [null],
>(endpoint: E, ...args: Args): UseErrorReturn<typeof args> {
  const adaptedEndpoint: any = useMemo(() => {
    return shapeToEndpoint(endpoint);
    // we currently don't support shape changes
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return useErrorNew(adaptedEndpoint, ...args);
}
