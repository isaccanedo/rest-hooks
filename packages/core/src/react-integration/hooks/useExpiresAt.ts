import { ReadShape, ParamsFromShape } from '../../endpoint/index.js';
import useMeta from './useMeta.js';

/** Returns whether the data at this url is fresh or stale
 * @deprecated use https://resthooks.io/docs/api/Controller#getResponse
 */
export default function useExpiresAt<
  Shape extends Pick<ReadShape<any, any>, 'getFetchKey' | 'options'>,
>(
  fetchShape: Shape,
  params: ParamsFromShape<Shape> | null,
  entitiesExpireAt = 0,
): number {
  const meta = useMeta(fetchShape, params);

  if (!meta) {
    return entitiesExpireAt;
  }

  return meta.expiresAt;
}
