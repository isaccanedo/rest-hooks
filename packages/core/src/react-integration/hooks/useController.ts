import { useContext } from 'react';

import Controller from '../../controller/Controller.js';
import { ControllerContext } from '../context.js';

/**
 * Imperative control of Rest Hooks store
 * @see https://resthooks.io/docs/api/useController
 */
export default function useController(): Controller {
  return useContext(ControllerContext);
}
