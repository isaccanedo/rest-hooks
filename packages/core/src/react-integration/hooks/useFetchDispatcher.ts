import { Schema } from '@rest-hooks/normalizr';
import { useContext, useCallback } from 'react';

import {
  FetchShape,
  SchemaFromShape,
  ParamsFromShape,
  BodyFromShape,
  OptimisticUpdateParams,
  ReturnFromShape,
} from '../../endpoint/index.js';
import { DispatchContext } from '../context.js';
import createFetch from '../../state/actions/createFetch.js';

/** Build an imperative dispatcher to issue network requests.
 * @deprecated use https://resthooks.io/docs/api/Controller#fetch
 */
export default function useFetchDispatcher(throttle = false): <
  Shape extends FetchShape<Schema, Readonly<object>, any>,
  UpdateParams extends OptimisticUpdateParams<
    SchemaFromShape<Shape>,
    FetchShape<any, any, any>
  >[],
>(
  fetchShape: Shape & {
    update?: (...args: any) => Record<string, (...args: any) => any>;
  },
  params: ParamsFromShape<Shape>,
  body: BodyFromShape<Shape>,
  updateParams?: UpdateParams | undefined,
) => ReturnFromShape<typeof fetchShape> {
  const dispatch = useContext(DispatchContext);

  const fetchDispatcher = useCallback(
    <Shape extends FetchShape<Schema, Readonly<object>, any>>(
      fetchShape: Shape,
      params: ParamsFromShape<Shape>,
      body: BodyFromShape<Shape>,
      updateParams?:
        | OptimisticUpdateParams<
            SchemaFromShape<Shape>,
            FetchShape<any, any, any>
          >[]
        | undefined,
    ) => {
      const action = createFetch(fetchShape, {
        params,
        body,
        throttle,
        updateParams,
      });
      dispatch(action);
      return action.meta.promise;
    },
    [dispatch, throttle],
  );
  // any is due to the ternary that we don't want to deal with in our implementation
  return fetchDispatcher as any;
}
