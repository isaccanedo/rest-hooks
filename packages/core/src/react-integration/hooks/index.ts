import useFetcher from './useFetcher.js';
import useController from './useController.js';
import useCache from './useCache.js';
import useRetrieve from './useRetrieve.js';
import useResource from './useResource.js';
import useSubscription from './useSubscription.js';
import useMeta from './useMeta.js';
import useError from './useError.js';
import useExpiresAt from './useExpiresAt.js';
import useInvalidator from './useInvalidator.js';
import useResetter from './useResetter.js';
import useFetchDispatcher from './useFetchDispatcher.js';
import useInvalidateDispatcher from './useInvalidateDispatcher.js';
import useFetch from '../newhooks/useFetch.js';
import useSuspense from '../newhooks/useSuspense.js';
export { default as hasUsableData } from './hasUsableData.js';
export type { ErrorTypes } from './useError.js';

export {
  useFetcher,
  useController,
  useFetchDispatcher,
  useCache,
  useError,
  useRetrieve,
  useResource,
  useSubscription,
  useMeta,
  useExpiresAt,
  useInvalidator,
  useInvalidateDispatcher,
  useResetter,
  useFetch,
  useSuspense,
};
