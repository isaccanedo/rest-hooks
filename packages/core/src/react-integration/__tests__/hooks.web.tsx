import {
  CoolerArticleResource,
  PaginatedArticleResource,
  ArticleResourceWithOtherListUrl,
  StaticArticleResource,
} from '__tests__/legacy-3';
import React, { Suspense, useEffect } from 'react';
import { render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import nock from 'nock';

// relative imports to avoid circular dependency in tsconfig references
import {
  makeRenderRestHook,
  makeCacheProvider,
  mockInitialState,
} from '../../../../test';
import { DispatchContext, StateContext } from '../context';
import { useFetcher, useRetrieve, useInvalidator, useResetter } from '../hooks';
import { initialState } from '../../state/createReducer';
import { State, ActionTypes } from '../../types';
import { INVALIDATE_TYPE, RESET_TYPE } from '../../actionTypes';
import { users, articlesPages, payload } from '../test-fixtures';

async function testDispatchFetch(
  Component: React.FunctionComponent<any>,
  payloads: any[],
) {
  const dispatch = jest.fn();
  const tree = (
    <DispatchContext.Provider value={dispatch}>
      <Suspense fallback={null}>
        <Component />
      </Suspense>
    </DispatchContext.Provider>
  );
  render(tree);
  expect(dispatch).toHaveBeenCalledTimes(payloads.length);
  let i = 0;
  for (const call of dispatch.mock.calls) {
    delete call[0]?.meta?.createdAt;
    delete call[0]?.meta?.promise;
    expect(call[0]).toMatchSnapshot();
    const action = call[0];
    const res = await action.payload();
    expect(res).toEqual(payloads[i]);
    i++;
  }
}

function testRestHook(
  callback: () => void,
  state: State<unknown>,
  dispatch = (v: ActionTypes) => Promise.resolve(),
) {
  return renderHook(callback, {
    wrapper: function Wrapper({ children }: { children: React.ReactNode }) {
      return (
        <DispatchContext.Provider value={dispatch}>
          <StateContext.Provider value={state}>
            {children}
          </StateContext.Provider>
        </DispatchContext.Provider>
      );
    },
  });
}

let mynock: nock.Scope;

beforeAll(() => {
  nock(/.*/)
    .persist()
    .defaultReplyHeaders({
      'Access-Control-Allow-Origin': '*',
      'Content-Type': 'application/json',
    })
    .options(/.*/)
    .reply(200);
  mynock = nock(/.*/).defaultReplyHeaders({
    'Access-Control-Allow-Origin': '*',
    'Content-Type': 'application/json',
  });
});

afterAll(() => {
  nock.cleanAll();
});

describe('useFetcher', () => {
  const payload = { id: 1, content: 'hi' };

  it('should dispatch an action that fetches a create', async () => {
    mynock.post(`/article-cooler/`).reply(201, payload);

    function DispatchTester() {
      const a = useFetcher(CoolerArticleResource.createShape());
      a({}, { content: 'hi' }).then(v => {
        v.author;
        //@ts-expect-error
        v.jasfdasdf;
      });
      return null;
    }
    await testDispatchFetch(DispatchTester, [payload]);
  });

  it('should dispatch an action with updater in the meta if update shapes params are passed in', async () => {
    mynock.post(`/article-cooler/`).reply(201, payload);

    function DispatchTester() {
      const create = useFetcher(CoolerArticleResource.createShape());
      const params = { content: 'hi' };
      create({}, params, [
        [
          CoolerArticleResource.listShape(),
          {},
          (article: any, articles: any) => [...articles, article],
        ],
      ]).then(v => {
        v.title;
        // @ts-expect-error
        v.doesnotexist;
      });
      return null;
    }
    await testDispatchFetch(DispatchTester, [payload]);
  });

  it('should refresh get details', async () => {
    mynock.get(`/article-cooler/${payload.id}`).reply(200, payload);

    function DispatchTester() {
      const refresh = useFetcher(CoolerArticleResource.detailShape());
      // @ts-expect-error
      () => refresh({}, { content: 'hi' });
      refresh({ id: payload.id }).then(v => {
        v.title;
        // @ts-expect-error
        v.doesnotexist;
      });
      return null;
    }
    await testDispatchFetch(DispatchTester, [payload]);
  });

  it('should dispatch an action with multiple updaters in the meta if update shapes params are passed in', async () => {
    mynock.post(`/article/`).reply(201, payload);

    function DispatchTester() {
      const create = useFetcher(ArticleResourceWithOtherListUrl.createShape());
      const params = { content: 'hi' };
      create({}, params, [
        [
          ArticleResourceWithOtherListUrl.listShape(),
          {},
          (article: any, articles: any) => [...articles, article],
        ],
        [
          ArticleResourceWithOtherListUrl.otherListShape(),
          {},
          (article: any, articles: any) => [...articles, article],
        ],
      ]);
      return null;
    }
    await testDispatchFetch(DispatchTester, [payload]);
  });

  it('should console.error() a warning when fetching without a Provider', () => {
    const oldError = console.error;
    const spy = (console.error = jest.fn());
    renderHook(() => {
      const a = useFetcher(CoolerArticleResource.createShape());
      a({}, { content: 'hi' });
      return null;
    });
    expect(spy.mock.calls[0]).toMatchInlineSnapshot(`
      [
        "It appears you are trying to use Rest Hooks without a provider.
      Follow instructions: https://resthooks.io/docs/getting-started/installation#add-provider-at-top-level-component",
      ]
    `);
    console.error = oldError;
  });

  it('should dispatch an action that fetches a partial update', async () => {
    mynock.patch(`/article-cooler/1`).reply(200, payload);

    function DispatchTester() {
      const a = useFetcher(CoolerArticleResource.partialUpdateShape());
      a({ id: payload.id }, { content: 'changed' });
      return null;
    }
    await testDispatchFetch(DispatchTester, [payload]);
  });

  it('should dispatch an action that fetches a full update', async () => {
    mynock.put(`/article-cooler/1`).reply(200, payload);

    function DispatchTester() {
      const a = useFetcher(CoolerArticleResource.updateShape());
      a({ id: payload.id }, { content: 'changed' });
      return null;
    }
    await testDispatchFetch(DispatchTester, [payload]);
  });
});

describe('useInvalidate', () => {
  it('should not invalidate anything if params is null', () => {
    const state = mockInitialState([
      {
        request: PaginatedArticleResource.listShape(),
        params: {},
        result: articlesPages,
      },
    ]);
    const dispatch = jest.fn();
    let invalidate: any;
    testRestHook(
      () => {
        invalidate = useInvalidator(PaginatedArticleResource.listShape());
      },
      state,
      dispatch,
    );
    invalidate(null);
    expect(dispatch).not.toHaveBeenCalled();
  });
  it('should return a function that dispatches an action to invalidate a resource', () => {
    const state = mockInitialState([
      {
        request: PaginatedArticleResource.listShape(),
        params: {},
        result: articlesPages,
      },
    ]);
    const dispatch = jest.fn();
    let invalidate: any;
    testRestHook(
      () => {
        invalidate = useInvalidator(PaginatedArticleResource.listShape());
      },
      state,
      dispatch,
    );
    invalidate({});
    expect(dispatch).toHaveBeenCalledWith({
      type: INVALIDATE_TYPE,
      meta: {
        key: 'GET http://test.com/article-paginated/',
      },
    });
  });
  it('should return the same === function each time', () => {
    const track = jest.fn();

    const { rerender } = renderHook(() => {
      const invalidate = useInvalidator(PaginatedArticleResource.listShape());
      useEffect(track, [invalidate]);
    });
    expect(track.mock.calls.length).toBe(1);
    for (let i = 0; i < 4; ++i) {
      rerender();
    }
    expect(track.mock.calls.length).toBe(1);
  });
});

describe('useResetter', () => {
  afterEach(() => {
    jest.useRealTimers();
  });
  it('should return a function that dispatches an action to reset the cache', () => {
    jest.useFakeTimers();
    const state = mockInitialState([
      {
        request: PaginatedArticleResource.listShape(),
        params: {},
        result: articlesPages,
      },
    ]);
    const dispatch = jest.fn();
    let reset: any;
    testRestHook(
      () => {
        reset = useResetter();
      },
      state,
      dispatch,
    );
    reset({});
    expect(dispatch).toHaveBeenCalledWith({
      type: RESET_TYPE,
      date: Date.now(),
    });
  });
  it('should return the same === function each time', () => {
    const track = jest.fn();

    const { rerender } = renderHook(() => {
      const reset = useResetter();
      useEffect(track, [reset]);
    });
    expect(track.mock.calls.length).toBe(1);
    for (let i = 0; i < 4; ++i) {
      rerender();
    }
    expect(track.mock.calls.length).toBe(1);
  });
});

describe('useRetrieve', () => {
  let renderRestHook: ReturnType<typeof makeRenderRestHook>;
  beforeEach(() => {
    mynock.get(`/article-cooler/${payload.id}`).reply(200, payload);
    mynock.get(`/article-static/${payload.id}`).reply(200, payload);
    mynock.get(`/user/`).reply(200, users);
    renderRestHook = makeRenderRestHook(makeCacheProvider);
  });
  afterEach(() => {
    nock.cleanAll();
  });

  it('should dispatch singles', async () => {
    function FetchTester() {
      useRetrieve(CoolerArticleResource.detailShape(), payload);
      return null;
    }
    await testDispatchFetch(FetchTester, [payload]);
  });

  it('should not dispatch will null params', () => {
    const dispatch = jest.fn();
    let params: any = null;
    const { rerender } = testRestHook(
      () => {
        useRetrieve(CoolerArticleResource.detailShape(), params);
      },
      initialState,
      dispatch,
    );
    expect(dispatch).toBeCalledTimes(0);
    params = payload;
    rerender();
    expect(dispatch).toBeCalled();
  });

  it('should dispatch with resource defined dataExpiryLength', async () => {
    function FetchTester() {
      useRetrieve(StaticArticleResource.detailShape(), payload);
      return null;
    }
    await testDispatchFetch(FetchTester, [payload]);
  });

  it('should dispatch with fetch shape defined dataExpiryLength', async () => {
    function FetchTester() {
      useRetrieve(StaticArticleResource.longLivingRequest(), payload);
      return null;
    }
    await testDispatchFetch(FetchTester, [payload]);
  });

  it('should dispatch with fetch shape defined errorExpiryLength', async () => {
    function FetchTester() {
      useRetrieve(StaticArticleResource.neverRetryOnErrorRequest(), payload);
      return null;
    }
    await testDispatchFetch(FetchTester, [payload]);
  });

  it('should not refetch after expiry and render', async () => {
    let time = 1000;
    global.Date.now = jest.fn(() => time);
    nock.cleanAll();
    const fetchMock = jest.fn(() => payload);
    mynock.get(`/article-cooler/${payload.id}`).reply(200, fetchMock).persist();
    const results: any[] = [
      {
        request: CoolerArticleResource.detailShape(),
        params: payload,
        result: payload,
      },
    ];
    const { result, rerender } = renderRestHook(
      () => {
        return useRetrieve(CoolerArticleResource.detailShape(), payload);
      },
      { results },
    );
    await result.current;
    expect(fetchMock).toHaveBeenCalledTimes(0);
    time += 100;
    rerender();
    await result.current;
    expect(fetchMock).toHaveBeenCalledTimes(0);
    // eslint-disable-next-line require-atomic-updates
    time += 610000000;
    rerender();
    await result.current;
    rerender();
    await result.current;
    expect(fetchMock).toHaveBeenCalledTimes(0);
  });
});
