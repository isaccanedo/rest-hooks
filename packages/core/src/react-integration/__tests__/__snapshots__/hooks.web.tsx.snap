// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useFetcher should dispatch an action that fetches a create 1`] = `
{
  "meta": {
    "args": [
      {},
      {
        "content": "hi",
      },
    ],
    "key": "POST http://test.com/article-cooler/",
    "options": {
      "errorPolicy": [Function],
      "fetchInit": {},
      "getFetchInit": [Function],
      "method": "POST",
      "signal": undefined,
      "url": [Function],
      "useFetchInit": [Function],
    },
    "reject": [Function],
    "resolve": [Function],
    "schema": [Function],
    "throttle": false,
    "type": "mutate",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useFetcher should dispatch an action that fetches a full update 1`] = `
{
  "meta": {
    "args": [
      {
        "id": 1,
      },
      {
        "content": "changed",
      },
    ],
    "key": "PUT http://test.com/article-cooler/1",
    "options": {
      "errorPolicy": [Function],
      "fetchInit": {},
      "getFetchInit": [Function],
      "method": "PUT",
      "signal": undefined,
      "url": [Function],
      "useFetchInit": [Function],
    },
    "reject": [Function],
    "resolve": [Function],
    "schema": [Function],
    "throttle": false,
    "type": "mutate",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useFetcher should dispatch an action that fetches a partial update 1`] = `
{
  "meta": {
    "args": [
      {
        "id": 1,
      },
      {
        "content": "changed",
      },
    ],
    "key": "PATCH http://test.com/article-cooler/1",
    "optimisticResponse": {
      "content": "changed",
      "id": 1,
    },
    "options": {
      "errorPolicy": [Function],
      "optimisticUpdate": [Function],
    },
    "reject": [Function],
    "resolve": [Function],
    "schema": [Function],
    "throttle": false,
    "type": "mutate",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useFetcher should dispatch an action with multiple updaters in the meta if update shapes params are passed in 1`] = `
{
  "meta": {
    "args": [
      {},
      {
        "content": "hi",
      },
    ],
    "key": "POST http://test.com/article/",
    "optimisticResponse": {
      "content": "hi",
    },
    "options": {
      "errorPolicy": [Function],
      "optimisticUpdate": [Function],
    },
    "reject": [Function],
    "resolve": [Function],
    "schema": [Function],
    "throttle": false,
    "type": "mutate",
    "update": [Function],
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useFetcher should dispatch an action with updater in the meta if update shapes params are passed in 1`] = `
{
  "meta": {
    "args": [
      {},
      {
        "content": "hi",
      },
    ],
    "key": "POST http://test.com/article-cooler/",
    "options": {
      "errorPolicy": [Function],
      "fetchInit": {},
      "getFetchInit": [Function],
      "method": "POST",
      "signal": undefined,
      "url": [Function],
      "useFetchInit": [Function],
    },
    "reject": [Function],
    "resolve": [Function],
    "schema": [Function],
    "throttle": false,
    "type": "mutate",
    "update": [Function],
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useFetcher should refresh get details 1`] = `
{
  "meta": {
    "args": [
      {
        "id": 1,
      },
      undefined,
    ],
    "key": "GET http://test.com/article-cooler/1",
    "options": {
      "errorPolicy": [Function],
      "fetchInit": {},
      "getFetchInit": [Function],
      "method": "GET",
      "signal": undefined,
      "url": [Function],
      "useFetchInit": [Function],
    },
    "reject": [Function],
    "resolve": [Function],
    "schema": [Function],
    "throttle": false,
    "type": "read",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useRetrieve should dispatch singles 1`] = `
{
  "meta": {
    "args": [
      {
        "content": "whatever",
        "id": 5,
        "tags": [
          "a",
          "best",
          "react",
        ],
        "title": "hi ho",
      },
      undefined,
    ],
    "key": "GET http://test.com/article-cooler/5",
    "options": {
      "errorPolicy": [Function],
      "fetchInit": {},
      "getFetchInit": [Function],
      "method": "GET",
      "signal": undefined,
      "url": [Function],
      "useFetchInit": [Function],
    },
    "reject": [Function],
    "resolve": [Function],
    "schema": [Function],
    "throttle": true,
    "type": "read",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useRetrieve should dispatch with fetch shape defined dataExpiryLength 1`] = `
{
  "meta": {
    "args": [
      {
        "content": "whatever",
        "id": 5,
        "tags": [
          "a",
          "best",
          "react",
        ],
        "title": "hi ho",
      },
      undefined,
    ],
    "key": "GET http://test.com/article-static/5",
    "options": {
      "dataExpiryLength": 3600000,
      "errorPolicy": [Function],
      "fetchInit": {},
      "getFetchInit": [Function],
      "method": "GET",
      "signal": undefined,
      "url": [Function],
      "useFetchInit": [Function],
    },
    "reject": [Function],
    "resolve": [Function],
    "schema": [Function],
    "throttle": true,
    "type": "read",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useRetrieve should dispatch with fetch shape defined errorExpiryLength 1`] = `
{
  "meta": {
    "args": [
      {
        "content": "whatever",
        "id": 5,
        "tags": [
          "a",
          "best",
          "react",
        ],
        "title": "hi ho",
      },
      undefined,
    ],
    "key": "GET http://test.com/article-static/5",
    "options": {
      "errorExpiryLength": Infinity,
      "errorPolicy": [Function],
      "fetchInit": {},
      "getFetchInit": [Function],
      "method": "GET",
      "signal": undefined,
      "url": [Function],
      "useFetchInit": [Function],
    },
    "reject": [Function],
    "resolve": [Function],
    "schema": [Function],
    "throttle": true,
    "type": "read",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;

exports[`useRetrieve should dispatch with resource defined dataExpiryLength 1`] = `
{
  "meta": {
    "args": [
      {
        "content": "whatever",
        "id": 5,
        "tags": [
          "a",
          "best",
          "react",
        ],
        "title": "hi ho",
      },
      undefined,
    ],
    "key": "GET http://test.com/article-static/5",
    "options": {
      "errorPolicy": [Function],
      "fetchInit": {},
      "getFetchInit": [Function],
      "method": "GET",
      "signal": undefined,
      "url": [Function],
      "useFetchInit": [Function],
    },
    "reject": [Function],
    "resolve": [Function],
    "schema": [Function],
    "throttle": true,
    "type": "read",
  },
  "payload": [Function],
  "type": "rest-hooks/fetch",
}
`;
